sql/sql_base.cc

```diff
@@ -1184,7 +1184,9 @@ TABLE_LIST* find_dup_table(THD *thd, TABLE_LIST *table, TABLE_LIST *table_list,
     DBUG_PRINT("info",
                ("found same copy of table or table which we should skip"));
   }
-  if (res && res->belong_to_derived)
+  // INSERT SELECT has better dealing with duplicates
+  if (res && res->belong_to_derived &&
+      !(check_flag & CHECK_DUP_FOR_INSERT_SELECT))
```

Comment:
Note the comment right below:

```
 We come here for queries of type:
      INSERT INTO t1 (SELECT tmp.a FROM (select * FROM t1) as tmp);
```
this is no longer true after this patch. The comment needs to be fixed.
(there will be more input).